mlflow/deployments/cli.py

```diff
@@ -0,0 +1,82 @@
+import click
+from mlflow.utils import cli_args
+from mlflow.deployments import interface
+from mlflow.deployments.utils import parse_custom_arguments
+
+
+deployment_target = click.option("--target", "-t", required=True, help="Deployment target")
+deployment_id = click.option("--id", "_deployment_id", required=True,
+                             help="Deployment ID for the deployment that needs to be deleted")
+context_settings = dict(allow_extra_args=True, ignore_unknown_options=True,)
+
+
+@click.group("deployments")
+def commands():
+    """
+    Deploy MLflow models. 
+    
+    MLflow will provide builtin support for some deployment targets eventually.
+    However, Support for ANY targets is only available via MLflow plugins right now - see
+    `community-plugins <https://mlflow.org/docs/latest/plugins.html#community-plugins>`_
+    for a list of supported plugins. MLFlow also enables you to write custom plugins for
+    deployment. For instructions on how to write and distribute your own plugin, checkout
+    `here <https://mlflow.org/docs/latest/plugins.html#writing-your-own-mlflow-plugins>`_.
+    """
+    pass
+
+
+@commands.command("create", context_settings=context_settings)
+@parse_custom_arguments
+@click.option("--flavor", "-f", help="Which flavor to be deployed. This will be auto "
+                                     "inferred if it's not given")
+@deployment_target
+@cli_args.MODEL_URI
+def create_cli(model_uri, target, flavor, **kwargs):
+    deployment = interface.create(target, model_uri, flavor, **kwargs)
+    # TODO: Support async here and everywhere requires
+    click.echo("\n{} deployment {} is created".format(deployment.flavor, deployment.id))
+
+
+@commands.command("delete", context_settings=context_settings)
+@parse_custom_arguments
+@deployment_id
+@deployment_target
+def delete_cli(target, _deployment_id, **kwargs):
+    interface.delete(target, _deployment_id, **kwargs)
+    click.echo("Deployment {} is deleted".format(_deployment_id))
+
+
+@commands.command("update", context_settings=context_settings)
+@parse_custom_arguments
+@click.option("--rollback", help="Should the deployment be rolled back. Rollback is an "
```

Comment:
Is there any disadvantage to just making `--rollback` a custom argument individual plugins can implement, rather than a standard argument that some plugins accept, but that other plugins handle by ignoring it or raising an exception?

Reply:
The API will be still confusing. For instance, we don't need the `model_uri`. So `model_uri` in the case of `rollback` and hence need to keep it as an optional parameter to the `update` command. This makes `--help` of `update` says, `model_uri` is optional (even for a non-rollback workflow) but the plugin throws an error if `model_uri` is not given.
I mean, the confusion here is not a deal-breaker, we can live with that. But we can live with the `--rollback` option being there too. Anyway, I am good with both, LMK what do you think

Reply:
Ah gotcha. What about making the API of `update` mirror the update of `create`, but with optional arguments? 
The idea would be that a deployment is defined by the tuple of arguments passed to create (`model_uri`, `flavor`, `**kwargs`) and that users should therefore be able to independently update one or more members of this tuple (although IMO `model_uri` should be required in order to specify a new `flavor` via `update`, it doesn't seem to add much value to be able to update the flavor independently of the `model_uri` and would make plugin logic more complex).

In that scenario, users could specify `model_uri` and `flavor` arguments, as well as custom `kwargs` defined by the plugin. If the user specifies none of these (or specifies `flavor` without `model_uri`), it makes sense for the `update` command to throw - otherwise, the update command should just update the desired fields.

This mirrors the pattern followed elsewhere in MLflow, e.g. in `update_registered_model`: https://github.com/mlflow/mlflow/blob/e0e19c6c19734b1847c417b3d59a799d9e007336/mlflow/tracking/client.py#L375

Reply:
IMO probably the best approach for supporting rollbacks (between adding it later as a separate command or keeping it as an argument to `update`) is just to leave them out entirely now, and think about how to support rollback functionality later on, especially as it's not super clear what the requirements around rollback functionality are for now (would be better to keep the interface simple IMO & collect user feedback on this).

If we do add rollback support later, it can be an API/CLI command separate from `update` that some plugins support, and for legacy plugins/plugins that don't implement `rollback`, users can always achieve a rollback by maintaining deployment history themselves & invoking `update` with appropriate arguments.

Reply:
We are going back to our initial approach here. Here are my thoughts
1. About rollback, I am OK to defer the implementation to get user feedback
2. About the signature of update; if that's the case, why would we even need the update command. We can have a `mode` argument (like in sagemaker's current implementation) and then we can minimize the surface

Reply:
Gotcha, yeah for 2) it still seems worthwhile to have an `update` command separate from `create`.

The difference between `update` here and in the sagemaker use case (where the CLI allows for `mlflow sagemaker deploy -m [replace/add/create]`  is that we want to be flexible to updating individual attributes of the deployment (e.g. for a model deployment on k8s we may want to specifically increase the number of model serving replicas). 

Moreover, the signatures of `create` and `update` are different - in the Sagemaker case, there's a common "app name" that defines a deployment & is shared across `create` and `update`, whereas in our case we allow the deployment ID to be defined by the deployment target, and don't know it at creation time

Reply:
Awesome, Thanks